core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -2026,6 +2026,40 @@ class SparkContext(config: SparkConf) extends Logging {
   }
 
   /**
+    * Create configs on driver using a provided function that accepts an integer as parameter
+    * and then run a function on a given set of partitions in an RDD and pass the results to
+    * the given handler function. This is the main entry point for all actions in Spark.
+    *
+    * @param rdd target RDD to run tasks on
+    * @param confFunc a function to create configurations on the driver
+    * @param func a function to run on each partition of the RDD
+    * @param partitions set of partitions to run on; some jobs may not want to compute on all
+    * partitions of the target RDD, e.g. for operations like `first()`
+    * @param resultHandler callback to pass each result to
+    */
+  private
+  def runJob[T, U: ClassTag](
```

Comment:
see my comments later on about refactoring & private apis --

in SparkContext, I'd just add one new `runJob`, which takes the extra function.  The existing methods should be changed to call that new one, with the a no-op function passed in.  That will simplify a lot of the other changes, as below this point you will know that extra function is *always* present

Reply:
but the runJob methods here are not private. I introduced new private ones to avoid changing apis

Reply:
yes, right we have to keep all the old methods here which are public.  You add one new which is private.  But all the others should change their *implementation* to call your new private version, so below this, they all have a "confFunc" (even if its a no-op)